tests/test_tuners_utils.py

```diff
@@ -415,6 +415,35 @@ def test_realistic_example(self):
         ]
         assert model.targeted_module_names == expected
 
+class TestExcludedModuleNames(unittest.TestCase):
```

Comment:
Can we get additional tests:

1. Let's add tests for `exclude_modules` being a list of str.
2. What happens if all targeted modules are excluded? I think a nice error message would be the right call in this situation.
3. What happens if we have excluded modules that don't exist? I think it's fine if some of the entries from `exclude_modules` don't match, as we also allow that for `target_modules`. But what if none of the entries in `exclude_modules` matches? I think this should also result in an error. WDYT?

Reply:
yeap makes sense, let me work through the tests and edge cases, also the missed out modules

Reply:
Hey @BenjaminBossan i managed to add 1 and 2 cases. I wanted some help on how 3 should be done. As in, I see currently _check_target_module_exists is a private abstract method that every peft methods override and check for target modules existence. Now for exclude_modules, should we do the same? Or do i open a for loop within inject_adapter and check for keys there?

Reply:
I don't think we need to go that far. If we check closer, we can see e.g. for LoRA:

https://github.com/huggingface/peft/blob/aa3bd8fbf6820bb000e148a289b6b0c48c339774/src/peft/tuners/lora/model.py#L158-L160

And if we check where that's used, it all comes back to this line:

https://github.com/huggingface/peft/blob/aa3bd8fbf6820bb000e148a289b6b0c48c339774/src/peft/tuners/tuners_utils.py#L486

So what would be an elegant way to implement some sanity checks here? I have one proposal, but maybe you have a better idea: Right now, `check_target_module_exists` simply returns a bool, i.e. `True` or `False` if the module matched. We could extend that and, say, return a special object if a module was excluded due to `exclude_modules`.

This special value could be a very simple object. Its `__bool__` should return `False` (so that users who rely on checking `if not self._check_target_module_exists(peft_config, key)` don't suddenly get different results).

When we call `_check_target_module_exists`, we can collect keys that did not match because of `exclude_modules` in one list and keys that did not match for other reasons in another list. When we exit the loop, if no key was matched because they were all matching `exclude_modules`, we can raise a nice error message to say so. Otherwise, we raise the existing error:

https://github.com/huggingface/peft/blob/aa3bd8fbf6820bb000e148a289b6b0c48c339774/src/peft/tuners/tuners_utils.py#L506-L510

On the other hand, if not a single key was excluded due to `exclude_modules`, even though `exclude_modules` was passed by the user, we can give a nice warning saying something like "You passed exclude_modules=[...] but no key was matching this".